### PR TITLE
ppx-deriving magic to improve code

### PR DESCRIPTION
**Pull request by smolkaj**
This pull request uses `[@@deriving ..]` to eliminate a bunch of boilerplate code, thus simplifying the code base. Changes include:

* Use `[@@deriving compare]` and  `[@@deriving eq]`to eliminate hand-coded comparison and equality functions. This eliminates an easy source of errors and makes the code more robust to changes.

* Use `[@@derving enum, enumerate]` to eliminate a nasty source of errors in the local compiler.

### CODE REVIEW

File: lib/Frenetic_Fdd.ml [hunk 1]
```diff
@@ -65,34 +66,24 @@ module Field = struct
     | Location -> "Location"
     | VFabric -> "VFabric"
```
Comment (smolkaj) [inline lib/Frenetic_Fdd.ml]: Adding or removing fields just got a lot easier! There are no more nasty dependencies between the definitions of `type t`, `num_fields`, and `all_fields`.

File: lib/Frenetic_Fdd.ml [hunk 2]
```diff
@@ -65,34 +66,24 @@ module Field = struct
...
-
   let is_valid_order (lst : t list) : bool =
-    List.length lst = num_fields &&
-    List.for_all all_fields ~f:(List.mem lst)
-
-  (* assert (is_valid_order all_fields) *)
+    Set.Poly.(equal (of_list lst) (of_list all))
 
   (* Initial order is the order in which fields appear in this file. *)
   let order = Array.init num_fields ~f:ident
```
Comment (smolkaj) [inline lib/Frenetic_Fdd.ml]: I was able to remove this imperative code. No more danger of having inconsistencies between `readable_order` and `order`.

**Comment by craig-riecke** [comment]
Sweet!  And it'll be nice when we flip over to OpenFlow 1.3 since that'll add a lot of fields to NetKAT.

**Comment by craig-riecke** [comment]
There's a small bug in this patch.  The Frenetic_Fdd.get_order function:

```
  let get_order () =
    Array.to_list order
    |> List.filter_map ~f:of_enum
```

is incorrect, which I found by automated tests created on the unit_tests branch.  get_order is currently only used in one place in the code, for a command line option default, so it's no biggie.  I'll correct it over in the unit_tests branch since I'm going to merge it in sometime soon.

**Comment by smolkaj** [comment]
Good catch!

**Pull request merged**